### ISSUE TITLE
fig up stuck if the container.wait throws exception

### ISSUE THREAD

**Issue by bigdrum** [docker/compose#953]
I use upstart and fig up to keep a bundle of containers running. If the any container exits, fig would exit and upstart would restart it. This works fine most of time. But occasionally (once per several day) fig would get stuck and do not exit.

I noticed the following exception at the end of the log when it stucked:

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/code/build/fig/out00-PYZ.pyz/threading", line 552, in __bootstrap_inner
  File "/code/build/fig/out00-PYZ.pyz/threading", line 505, in run
  File "/code/build/fig/out00-PYZ.pyz/fig.cli.multiplexer", line 41, in _enqueue_output
  File "/code/build/fig/out00-PYZ.pyz/fig.cli.log_printer", line 59, in _make_log_generator
  File "/code/build/fig/out00-PYZ.pyz/fig.container", line 141, in wait
  File "/code/build/fig/out00-PYZ.pyz/docker.client", line 940, in wait
  File "/code/build/fig/out00-PYZ.pyz/docker.client", line 73, in _post
  File "/code/build/fig/out00-PYZ.pyz/requests.sessions", line 425, in post
  File "/code/build/fig/out00-PYZ.pyz/requests.sessions", line 383, in request
  File "/code/build/fig/out00-PYZ.pyz/requests.sessions", line 486, in send
  File "/code/build/fig/out00-PYZ.pyz/requests.adapters", line 378, in send
ConnectionError: UnixHTTPConnectionPool(host='localhost', port=None): Max retries exceeded with url: /run/docker.sock/v1.12/containers/55b10e13d8449801bfac24d58134f7a71a6f61412021e104369e776a2039872e/wait (Caused by <class 'socket.error'>: [Errno 111] Connection refused)
```

afaick, if container.wait() throws, fig just crashes the thread and leaves the multiplexer waiting on the queue forever. And fig won't exit if there are containers that are still doing fine.

It seems the log printer should send insert STOP to the queue when there is any uncaught exception thrown in the thread.

**Comment by mishak87** [comment]
I am getting it regularly while machine is doing setup without any output like installing packages or sleep infinity.
```
Exception in thread Thread-9:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/compose/cli/multiplexer.py", line 41, in _enqueue_output
    for item in generator:
  File "/usr/local/lib/python2.7/dist-packages/compose/cli/log_printer.py", line 59, in _make_log_generator
    for line in line_generator:
  File "/usr/local/lib/python2.7/dist-packages/compose/cli/utils.py", line 100, in split_buffer
    for data in reader:
  File "/usr/local/lib/python2.7/dist-packages/docker/clientbase.py", line 238, in _stream_raw_result
    for out in response.iter_content(chunk_size=1, decode_unicode=True):
  File "/usr/local/lib/python2.7/dist-packages/requests/utils.py", line 332, in stream_decode_response_unicode
    for item in iterator:
  File "/usr/local/lib/python2.7/dist-packages/requests/models.py", line 680, in generate
    raise ConnectionError(e)
ConnectionError: UnixHTTPConnectionPool(host='localhost', port=None): Read timed out.
```

**Comment by Yajo** [comment]
Same for me. Seems to happen when an image has `tty: true` in its configuration.

```
> docker-compose version
docker-compose version: 1.4.2
docker-py version: 1.3.1
CPython version: 2.7.10
OpenSSL version: OpenSSL 1.0.2e-fips 3 Dec 2015
```

**Comment by dreyks** [comment]
I can confirm on `tty: true` causing this.
Actually I _do_ want a tty allocated so this is a not so minor issue for me.

**Comment by ijc** [comment]
Issue grooming: It's been a long time since the last report, so I am taking a guess that this has been fixed in the meantime. If not and you are seeing this with a current version then please open a new issue with reproduction steps.
